### PR TITLE
berlin

### PR DESCRIPTION
**Pull request by MrChico**
Superseeds #634

### CODE REVIEW

**Comment by MrChico** [comment]
so it begins d875cd07ebfb8c6433ad8eaed9770a9644cf7c88
Screen Shot 2021-04-21 at 19 51 56

**Comment by MrChico** [comment]
5b4138d0a939903375ecb1a7b091118a150f584c:
```
passed: 7821
bad-balance: 4130
bad-nonce: 0
bad-storage: 3
failed: 194
timeout: 5
skipped: 0
```

**Comment by MrChico** [comment]
f40635faaaa70449af8ce4b6fe357958070322a7:
Screen Shot 2021-04-21 at 21 01 19

**Comment by laudiacay** [comment]
my machine is passing two more of the timeouts (ie, 3 timeout, 11857 pass). working on these now though :)

File: src/hevm/src/EVM.hs [hunk 1]
```diff
@@ -832,8 +850,10 @@ exec1 = do
               : codeSize'
               : xs ) ->
               forceConcrete4 (extAccount', memOffset', codeOffset', codeSize') $
-                \(extAccount, memOffset, codeOffset, codeSize) ->
-                  burn (g_extcode + g_copy * ceilDiv (num codeSize) 32) $
+                \(extAccount, memOffset, codeOffset, codeSize) -> do
+                  acc <- accessAccountForGas (num extAccount)
+                  let cost = if acc then g_warm_storage_read else g_cold_account_access
+                  burn cost $
```
Comment (MrChico) [inline src/hevm/src/EVM.hs:856]: hmmm

**Comment by MrChico** [comment]
0dc21f00802b90d9424c50d0cbdf3265b7b9ff7d:
Screen Shot 2021-04-23 at 05 00 37

**Comment by MrChico** [comment]
18484a6d5f7285ba6dc1fdd8835621758b230506:
Screen Shot 2021-04-23 at 06 18 45

**Comment by MrChico** [comment]
@laudiacay looks like there are still some failures in the modexp calculation: `hevm bc-test --file ./BlockchainTests/GeneralStateTests/stPreCompiledContracts2/modexpRandomInput.json` gives some failures. Since I imagine that's still pretty fresh in your head I'll let you dig into that while I keep tracking down these other issues

**Comment by MrChico** [comment]
cdb21603ea6dbf674644ed97c3db47ee02075601:
Screen Shot 2021-04-23 at 17 19 19

**Comment by MrChico** [comment]
do not understand whats going on with BlockchainTests/GeneralStateTests/stExtCodeHash/extcodehashEmpty.json

**Comment by MrChico** [comment]
1f4f6cb546904d74da5d53391d264b5de2548428:
Screen Shot 2021-04-26 at 18 24 25

**Comment by MrChico** [comment]
Now running VMTests as well, found issues in addmod and mulmod.
c89ba24e13c0fe824c0cb0144d220410cfed9fd8:
Screen Shot 2021-04-27 at 19 57 39

File: src/hevm/src/EVM.hs [hunk 2]
```diff
@@ -1339,7 +1368,6 @@ precompiledContract this xGas precompileAddr recipient xValue inOffset inSize ou
           transfer self recipient xValue
           touchAccount self
           touchAccount recipient
-          touchAccount precompileAddr
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:1342]: why?

File: src/hevm/src/EVM.hs [hunk 3]
```diff
@@ -2693,6 +2751,20 @@ costOfCreate (FeeSchedule {..}) availableGas' hashSize =
     hashCost   = g_sha3word * ceilDiv (num hashSize) 32
     initGas    = allButOne64th (availableGas - createCost)
 
+concreteModexpGasFee :: ByteString -> Integer
+concreteModexpGasFee input = num $ max 200 ((multiplicationComplexity * iterCount) `div` 3)
+  where (lenb, lene, lenm) = parseModexpLength input
+        ez = isZero (96 + lenb) lene input
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:2757]: where does the `96` come from?

File: src/hevm/src/EVM.hs [hunk 4]
```diff
@@ -2693,6 +2751,20 @@ costOfCreate (FeeSchedule {..}) availableGas' hashSize =
...
+        e' = w256 $ word $ LS.toStrict $
+          lazySlice (96 + lenb) (min 32 lene) input
+        nwords = ceiling $ num (max lenb lenm) / 8
+        multiplicationComplexity = nwords * nwords
+        iterCount' | lene <= 32 && ez = 0
+                   | lene <= 32 = num (log2 e')
+                   | e' == 0 = 8 * (lene - 32)
+                   | otherwise = num (log2 e') + 8 * (lene - 32)
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:2765]: I don't understand these two cases.

In the [EIP](https://eips.ethereum.org/EIPS/eip-2565#specification) I see the following for the final case, which doesn't seem to map onto the code here?

```python
(8 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
```

File: src/hevm/src/EVM.hs [hunk 5]
```diff
@@ -2706,24 +2778,10 @@ costOfPrecompile (FeeSchedule {..}) precompileAddr input =
...
-
-            ez = isZero (96 + lenb) lene input'
-            e' = w256 $ word $ LS.toStrict $
-                   lazySlice (96 + lenb) (min 32 lene) input'
-
-            f :: Integer -> Integer
-            f x | x <= 64 = x * x
-                | x <= 1024 = (x * x) `div` 4 + 96 * x - 3072
-                | otherwise = (x * x) `div` 16 + 480 * x - 199680
+               SymbolicBuffer _ -> error "unsupported: symbolic MODEXP gas cost calc"
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM.hs:2783]: can't we fail properly here with an `UnexpectedSymbolicArg`?

File: src/hevm/src/EVM/Concrete.hs [hunk 6]
```diff
@@ -105,7 +105,7 @@ x0 ^ y0 | y0 < 0    = errorWithoutStackTrace "Negative exponent"
                   | otherwise   = g (x * x) ((y - 1) `shiftR` 1) (x * z)
 
 createAddress :: Addr -> W256 -> Addr
-createAddress a n = num $ keccak $ rlpList [rlpWord160 a, rlpWord256 n]
+createAddress a n = num $ keccak $ rlpList [rlpAddrFull a, rlpWord256 n]
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Concrete.hs:108]: what is the reason for the change here?

File: src/hevm/src/EVM/Exec.hs [hunk 7]
```diff
@@ -36,10 +36,11 @@ vmForEthrunCreation creationCode =
     , vmoptGas = 0xffffffffffffffff
     , vmoptGaslimit = 0xffffffffffffffff
     , vmoptMaxCodeSize = 0xffffffff
-    , vmoptSchedule = FeeSchedule.istanbul
+    , vmoptSchedule = FeeSchedule.berlin
     , vmoptChainId = 1
     , vmoptCreate = False
     , vmoptStorageModel = ConcreteS
+    , vmoptTxAccessList = mempty  -- TODO is this right
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Exec.hs:43]: what else would it be?

File: src/hevm/src/EVM/FeeSchedule.hs [hunk 8]
```diff
@@ -158,3 +168,18 @@ eip2200 fees = fees
...
+  { g_sload = 100 
+  , g_sreset = 5000 - 2100 
+  , g_call = 2600
+  , g_balance = 2600
+  , g_extcode = 2600
+  , g_extcodehash = 2600
+  }
+
+berlin :: Num n => FeeSchedule n
+berlin = eip2929 istanbul
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/FeeSchedule.hs:185]: total nit, but there is a missing newline here...

File: src/hevm/src/EVM/Types.hs [hunk 9]
```diff
@@ -364,7 +364,7 @@ instance Show Addr where
   showsPrec _ addr next =
     let hex = showHex addr next
         str = replicate (40 - length hex) '0' ++ hex
-    in "0x" ++ toChecksumAddress str
+    in "0x" ++ toChecksumAddress str ++ drop 40 str
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Types.hs:367]: what is this for?

File: src/hevm/src/EVM/Transaction.hs [hunk 10]
```diff
@@ -140,11 +202,6 @@ initTx vm = let
...
       $ preState
 
-    touched = if creation
-              then [origin]
-              else [origin, toAddr]
-
     in
       vm & EVM.env . EVM.contracts .~ initState
          & EVM.tx . EVM.txReversion .~ preState
-         & EVM.tx . EVM.substate . EVM.touchedAccounts .~ touched
```
Comment (xwvvvvwx) [inline src/hevm/src/EVM/Transaction.hs:150]: why?

File: src/hevm/src/EVM/Concrete.hs (hunk 6 above)
Comment (MrChico) [reply src/hevm/src/EVM/Concrete.hs:108]: `rlpWord160` expresses addresses as numbers, in the sense that `0x0000000000000000000000007265636569766572` gets encoded without leading zeros. This gives the wrong creation address

File: src/dapp-tests/default.nix [hunk 11]
```diff
@@ -142,7 +142,7 @@ in
     dss = pkgs.buildDappPackage {
       src = dss-src;
       name = "dss";
-      doCheck = true;
+      doCheck = false; # TODO: switch me on when dai tests are updated.
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/dapp-tests/default.nix:145]: The dai tests are failing because the address changed with 1f4f6cb546904d74da5d53391d264b5de2548428

File: src/hevm/src/EVM/FeeSchedule.hs (hunk 8 above)
Comment (MrChico) [inline src/hevm/src/EVM/FeeSchedule.hs:185]:
```suggestion
berlin = eip2929 istanbul

```

File: src/hevm/src/EVM/Transaction.hs (hunk 10 above)
Comment (MrChico) [reply src/hevm/src/EVM/Transaction.hs:150]: moved to `EVM.makeVm`

File: src/hevm/src/EVM/Types.hs (hunk 9 above)
Comment (MrChico) [reply src/hevm/src/EVM/Types.hs:367]: `next` is the continuation of things that need to be printed. Not keeping it means printing an address can eat up the rest of the things you wanna print

**Comment by MrChico** [comment]
6059b47ec13e82c15bfaeec2d8c7a10781248db5:
Screen Shot 2021-04-28 at 20 16 57

File: src/hevm/src/EVM/Exec.hs (hunk 7 above)
Comment (MrChico) [inline src/hevm/src/EVM/Exec.hs:43]:
```suggestion
    , vmoptTxAccessList = mempty
```

File: src/hevm/src/EVM.hs [hunk 12]
```diff
@@ -731,8 +745,10 @@ exec1 = do
         -- op: BALANCE
         0x31 ->
           case stk of
-            (x':xs) -> forceConcrete x' $ \x ->
-              burn g_balance $
+            (x':xs) -> forceConcrete x' $ \x -> do
+              acc <- accessAccountForGas (num x)
+              let cost = if acc then g_warm_storage_read else g_cold_account_access
+              burn cost $
```
[the remark this answers is not in the document]
Comment (MrChico) [reply src/hevm/src/EVM.hs:751]: fixed

File: src/hevm/src/EVM.hs (hunk 3 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:2757]: input is given on the following form:
`<length_of_BASE> <length_of_EXPONENT> <length_of_MODULUS> <BASE> <EXPONENT> <MODULUS>`
the first three are all 32 bytes. therefore `<BASE>` begins at 96

File: src/hevm/src/EVM.hs (hunk 4 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs]: exponent_length is `lene`, `x.bit_length - 1` is given by `log2 x`, the truncation to a 256 word happens before in `min 32 lene` on L2702

File: src/hevm/src/EVM.hs (hunk 5 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:2785]: not without changing the type signature here

File: src/hevm/src/EVM.hs (hunk 2 above)
Comment (MrChico) [reply src/hevm/src/EVM.hs:1342]: DELEGATECALL and CALLCODE doesn't touch the address they fetch the code from

Comment (MrChico) [reply src/hevm/src/EVM.hs:1342]: I don't know a better explanation to "why" than "geth does this"

File: release.nix [hunk 13]
```diff
@@ -20,7 +20,7 @@ let
     export PATH=${x.pkgs.hevm}/bin:${x.pkgs.jq}/bin:$PATH
     ${x.pkgs.hevm}/bin/hevm compliance \
       --tests ${ethereum-test-suite x} \
-      --skip "(RevertPrecompiledTouch_storage|Create2Recursive|Create1000|recursiveCreateReturn|randomStatetest647|multiOwnedRemoveOwner|walletRemoveOwnerRemovePending|static_Call50000|Return5000)" \
+      --skip "(Create2Recursive|Create1000|recursiveCreateReturn|underflowTest|static_Call50000|Return5000|randomStatetest177|loopExp|loopMul)" \
```
[the remark this answers is not in the document]
Comment (MrChico) [reply release.nix]: fleshed out the comment that was already there. All of them are timeouts or nonstandard format except for `walletRemoveOwnerRemovePending` which is unfortunately still a consensus failure for hevm. There is an issue tracking this at #328 but I think we should not let that block this PR any further

**Comment by MrChico** [comment]
all tests passing except https://github.com/dapphub/dapptools/pull/636#discussion_r622609778, good to merge imo

**Review by xwvvvvwx — APPROVED** [review]
nice!!! :sparkling_heart:
